MdeModulePkg/Library/MemoryBucketLib/MemoryBucketLib.c

```diff
@@ -0,0 +1,447 @@
+
+// TRUE if we have initialized the runtime memory buckets.
+BOOLEAN  RuntimeMemInitialized = FALSE;
+
+/**
+  TODO
+**/
+VOID
+EFIAPI
+InitializeMemoryBucketSizes (
+  VOID
+  )
+{
+  RuntimeMemoryStats[0].NumberOfPages = FixedPcdGet8 (PcdPeiMemoryBucketRuntimeCode);
+  RuntimeMemoryStats[1].NumberOfPages = FixedPcdGet8 (PcdPeiMemoryBucketRuntimeData);
+  RuntimeMemoryStats[2].NumberOfPages = FixedPcdGet8 (PcdPeiMemoryBucketAcpiReclaimMemory);
+  RuntimeMemoryStats[3].NumberOfPages = FixedPcdGet8 (PcdPeiMemoryBucketAcpiMemoryNvs);
+
+  // Disable memory buckets if the PCDs are unaltered.
+  if ((RuntimeMemoryStats[0].NumberOfPages + RuntimeMemoryStats[1].NumberOfPages +
+       RuntimeMemoryStats[2].NumberOfPages + RuntimeMemoryStats[3].NumberOfPages) == 0)
+  {
+    MemoryBucketsDisabled = TRUE;
+  }
+}
+
+/**
+  This function initialized the PEI memory buckets.  This should be called
+  when the first memory type being tracked in these buckets is allocated.
+
+  NOTE: The memory buckets can be created in Pre-mem PEI if runtime pages are
+  allocated in pre-mem PEI.
+
+  @param[in] StartingAddress   The starting address of the memory buckets.
+                               All of them are contiguous.
+**/
+VOID
+EFIAPI
+InitializeMemoryBuckets (
+  EFI_PHYSICAL_ADDRESS  StartingAddress
+  )
+{
+  UINTN   Index;
+  UINTN   AdjustedIndex;
+  UINT32  AddressAdjustment;
+
+  InitializeMemoryBucketSizes ();
+  AdjustedIndex = 0;
+
+  for (Index = 0; Index < NumberOfBuckets; Index++) {
+    // Initialize memory locations for buckets
+    AddressAdjustment = EFI_PAGE_SIZE *
+                        ((UINT32)RuntimeMemoryStats[Index].NumberOfPages * (UINT32)AdjustedIndex);
+    RuntimeMemoryStats[Index].BaseAddress = StartingAddress - AddressAdjustment;
+    CurrentBucketTops[Index]              = StartingAddress - AddressAdjustment;
+
+    // Initialize Hob to be able to reference it later
+    RuntimeBucketHob.RuntimeBuckets[Index]     = RuntimeMemoryStats[Index];
+    RuntimeBucketHob.CurrentTopInBucket[Index] = CurrentBucketTops[Index];
+    if (RuntimeMemoryStats[Index].NumberOfPages != 0) {
+      AdjustedIndex += 1;
+    }
+  }
+
+  RuntimeBucketHob.MemoryBucketsDisabled = MemoryBucketsDisabled;
+}
+
+/**
+  This function translates the memory type being allocated into the
+  index that it uses in the memory buckets structure.
+
+  @param[in] MemoryType   The type of memory we are interested in.
+
+  @retval    The index associated with the memory type.
+**/
+UINTN
+EFIAPI
+MemoryTypeToIndex (
+  EFI_MEMORY_TYPE  MemoryType
+  )
+{
+  UINTN  Index;
+
+  switch (MemoryType) {
+    case EfiRuntimeServicesCode:
+      Index = 0;
+      break;
+    case EfiRuntimeServicesData:
+      Index = 1;
+      break;
+    case EfiACPIReclaimMemory:
+      Index = 2;
+      break;
+    case EfiACPIMemoryNVS:
+      Index = 3;
+      break;
+    default:
+      DEBUG ((DEBUG_ERROR, "We got an incorrect MemoryType\n"));

```

Comment:
Recommend adding function name. Will make a lot more sense in the context of the full debug log.